lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3918,74 +3919,35 @@ void AttributeChecker::visitDerivativeAttr(DerivativeAttr *attr) {
     return;
   }
 
-  // Try to find a `@differentiable` attribute on the original function with the
-  // same differentiation parameters.
-  DifferentiableAttr *da = nullptr;
-  for (auto *cda : originalAFD->getAttrs().getAttributes<DifferentiableAttr>())
-    if (checkedWrtParamIndices == cda->getParameterIndices())
-      da = const_cast<DifferentiableAttr *>(cda);
-  // If the original function does not have a `@differentiable` attribute with
-  // the same differentiation parameters, create one.
-  // TODO(TF-835): Lower `@derivative` attributes directly to SIL
-  // differentiability witnesses during SILGen instead of generating implicit
-  // `@differentiable` attributes.
-  if (!da) {
-    da = DifferentiableAttr::create(
-        originalAFD, /*implicit*/ true, attr->AtLoc, attr->getRange(),
-        /*linear*/ false, checkedWrtParamIndices, /*jvp*/ None,
-        /*vjp*/ None, derivative->getGenericSignature());
-    switch (kind) {
-    case AutoDiffDerivativeFunctionKind::JVP:
-      da->setJVPFunction(derivative);
-      break;
-    case AutoDiffDerivativeFunctionKind::VJP:
-      da->setVJPFunction(derivative);
-      break;
-    }
-    auto insertion = Ctx.DifferentiableAttrs.try_emplace(
-        {originalAFD, checkedWrtParamIndices}, da);
-    // Valid `@differentiable` attributes are uniqued by their parameter
-    // indices. Reject duplicate attributes for the same decl and parameter
-    // indices pair.
-    if (!insertion.second && insertion.first->getSecond() != da) {
-      diagnoseAndRemoveAttr(da, diag::differentiable_attr_duplicate);
-      diagnose(insertion.first->getSecond()->getLocation(),
-               diag::differentiable_attr_duplicate_note);
-      return;
-    }
-    originalAFD->getAttrs().add(da);
-    return;
+  // Diagnose if there exists a `@differentiable` attribute with the same
+  // parameter indices as the `@derivative` function.
+  // NOTE(TF-680): relaxing this limitation requires derivative generic
+  // signature mangling to avoid name collisions for SIL derivative functions
+  // with the same parameter indices but different derivative generic
+  // signatures.
+  auto *diffAttr =
+      Ctx.DifferentiableAttrs.lookup({originalAFD, checkedWrtParamIndices});
+  bool isOriginalAFDProtocolRequirement =
+      isa<ProtocolDecl>(originalAFD->getDeclContext()) &&
+      originalAFD->isProtocolRequirement();
+  if (diffAttr && !isOriginalAFDProtocolRequirement) {
+    diagnoseAndRemoveAttr(attr,
+                          diag::derivative_attr_original_already_has_derivative,
```

Comment:
I reused the `derivative_attr_original_already_has_derivative ` diagnostic, but it arguably has room for improvement:
```
/swift-base/tensorflow-swift-apis/Sources/TensorFlow/Layer.swift:94:6: error: a derivative already exists for 'inferring(from:)'
    @derivative(of: inferring(from:))
    ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    
/swift-base/tensorflow-swift-apis/Sources/TensorFlow/Layer.swift:88:6: note: other attribute declared here
    @differentiable
     ^
```
Showing `"a derivative already exists"` for a `@differentiable` attribute may not be clear to users who don't know that `@differentiable` attribute triggers derivative function generation.

I'm open to improvement suggestions, if anyone has ideas. Otherwise, I'll defer 'til later.

Reply:
Obsolete because this diagnostic is no longer being emitted. `@differentiable` + `@derivative` attributes are never in conflict: https://github.com/apple/swift/pull/28621#issuecomment-562751760